src/java/org/apache/cassandra/cql3/CqlConstraint.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.schema.Types;
+import org.apache.cassandra.schema.UserFunctions;
+import org.apache.cassandra.tcm.serialization.Version;
+
+public class CqlConstraint
+{
+    public ColumnIdentifier constraintName;
+    public final ColumnIdentifier columnName;
+    public final ConstraintCondition constraintCondition;
+
+    public static Serializer serializer = new Serializer();
+
+    public final static class Raw {
+        public final ColumnIdentifier constraintName;
+        public final ConstraintCondition constraintCondition;
+        public ColumnIdentifier columnName;
+
+        public Raw(ColumnIdentifier constraintName, ConstraintCondition constraintCondition)
+        {
+            this.constraintName = constraintName;
+            this.constraintCondition = constraintCondition;
+        }
+
+        public CqlConstraint prepare(ColumnIdentifier columnName)
+        {
+            return new CqlConstraint(constraintName, columnName, constraintCondition);
+        }
+
+        public CqlConstraint prepareWithName(ColumnIdentifier constraintName)
+        {
+            return new CqlConstraint(constraintName, null, constraintCondition);
+        }
+    }
+
+    public CqlConstraint(ColumnIdentifier constraintName, ColumnIdentifier columnName, ConstraintCondition constraintCondition)
+    {
+        if (constraintName == null)
+        {
+            final String randomConstraintName = UUID.randomUUID().toString().replace("-", "");
+            this.constraintName = new ColumnIdentifier(randomConstraintName, false);
+        }
+        else
+        {
+            this.constraintName = constraintName;
+        }
+        this.columnName = columnName;
+        this.constraintCondition = constraintCondition;
+    }
+
+    public void appendCqlTo(CqlBuilder builder)
+    {
+        builder.append(" CHECK ").append(toString());
+    }
+
+    public void checkConstraint(Map<String, String> columnValues, ColumnMetadata columnMetadata, TableMetadata tableMetadata)
+    {
+        constraintCondition.checkCondition(columnValues, columnMetadata, tableMetadata);
+    }
+
+    public void validateConstraint(ColumnMetadata columnMetadata, TableMetadata tableMetadata)
+    {
+        constraintCondition.validateCondition(columnMetadata, tableMetadata);
+    }
+
+    @Override
+    public String toString()
+    {
+        return constraintCondition.toString();
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hashCode(constraintName);
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        return obj instanceof CqlConstraint
+               && Objects.equal(constraintName, ((CqlConstraint) obj).constraintName)
+               && Objects.equal(constraintCondition, ((CqlConstraint) obj).constraintCondition);
```

Comment:
Missing `obj == this` case?

Reply:
Wouldn't that be included in the current return?

Reply:
`obj == this` is a shortcut to avoid comparing values. It exits as soon as finding out the object reference are the same, i.e. same object.